File: server/src/modules/system/menu/menu.controller.ts
import { Context } from "koa";
import menuService from "./menu.service";
import { Result } from "@/core/result";

class MenuController {
  // GET /menu/list — 完整菜单树（管理端），支持 title 筛选
  async list(ctx: Context) {
    const { title } = ctx.query as any;
    const res = await menuService.listAll(title);
    ctx.body = res;
  }

  // GET /menu/tree — 当前用户的导航菜单树（动态路由专用）
  async tree(ctx: Context) {
    const { id } = ctx.state.user;
    const res = await menuService.getUserMenus(id);
    ctx.body = res;
  }

  // POST /menu/add
  async add(ctx: Context) {
    const data = ctx.request.body as any;
    if (!data.title) return (ctx.body = Result.error("菜单名称不能为空"));

    if (data.roles && Array.isArray(data.roles)) {
      data.roles = JSON.stringify(data.roles);
    }
    if (data.authList && Array.isArray(data.authList)) {
      data.authList = JSON.stringify(data.authList);
    }

    const res = await menuService.add(data);
    ctx.body = Result.success(res);
  }

  // POST /menu/update
  async update(ctx: Context) {
    const { id, ...data } = ctx.request.body as any;
    if (!id) return (ctx.body = Result.error("ID不能为空"));

    if (data.roles && Array.isArray(data.roles)) {
      data.roles = JSON.stringify(data.roles);
    }
    if (data.authList && Array.isArray(data.authList)) {
      data.authList = JSON.stringify(data.authList);
    }

    const res = await menuService.update(id, data);
    ctx.body = Result.success(res);
  }

  // POST /menu/delete
  async delete(ctx: Context) {
    const { id } = ctx.request.body as any;
    if (!id) return (ctx.body = Result.error("ID不能为空"));

    const hasChildren = await (menuService as any).model.findFirst({
      where: { parentId: id },
    });
    if (hasChildren) return (ctx.body = Result.error("存在子菜单，请先删除子菜单"));

    await menuService.delete(id);
    ctx.body = Result.success(null, "删除成功");
  }
}

export default new MenuController();

File: java-server/src/main/java/com/todothink/modules/system/menu/controller/MenuController.java
package com.todothink.modules.system.menu.controller;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.todothink.core.result.Result;
import com.todothink.core.security.SecurityUtils;
import com.todothink.modules.system.entity.Menu;
import com.todothink.modules.system.menu.service.MenuService;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/menu")
public class MenuController {

    private final MenuService menuService;
    private final ObjectMapper objectMapper;

    public MenuController(MenuService menuService, ObjectMapper objectMapper) {
        this.menuService = menuService;
        this.objectMapper = objectMapper;
    }

    @GetMapping("/list")
    public Result<List<Map<String, Object>>> list(@RequestParam(required = false) String title) {
        return Result.success(menuService.listAll(title));
    }

    @GetMapping("/tree")
    public Result<List<Map<String, Object>>> tree() {
        return Result.success(menuService.getUserMenus(SecurityUtils.currentUserId()));
    }

    @PostMapping("/add")
    public Result<Menu> add(@RequestBody Map<String, Object> body) {
        if (body.get("title") == null) {
            return Result.error("菜单名称不能为空");
        }
        Menu menu = objectMapper.convertValue(body, Menu.class);
        normalizeJsonFields(menu, body);
        return Result.success(menuService.add(menu));
    }

    @PostMapping("/update")
    public Result<Menu> update(@RequestBody Map<String, Object> body) {
        Object idObj = body.get("id");
        if (idObj == null) {
            return Result.error("ID不能为空");
        }
        Integer id = Integer.valueOf(String.valueOf(idObj));
        Menu menu = objectMapper.convertValue(body, Menu.class);
        normalizeJsonFields(menu, body);
        return Result.success(menuService.update(id, menu));
    }

    @PostMapping("/delete")
    public Result<Void> delete(@RequestBody Map<String, Object> body) {
        Object idObj = body.get("id");
        if (idObj == null) {
            return Result.error("ID不能为空");
        }
        menuService.delete(Integer.valueOf(String.valueOf(idObj)));
        return Result.success(null, "删除成功");
    }

    private void normalizeJsonFields(Menu menu, Map<String, Object> body) {
        try {
            if (body.get("roles") instanceof List) {
                menu.setRoles(objectMapper.writeValueAsString(body.get("roles")));
            }
            if (body.get("authList") instanceof List) {
                menu.setAuthList(objectMapper.writeValueAsString(body.get("authList")));
            }
        } catch (Exception ignored) {
            // ignore
        }
    }
}
